src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala

```diff
@@ -97,28 +99,51 @@ trait WorkspaceComponent {
+        case None =>
+          (workspaceQuery += marshalNewWorkspace(workspace)) andThen
+            insertOrUpdateAccessRecords(workspace) andThen
+            insertAttributes(workspace)
+        case Some(workspaceRecord) =>
+          insertOrUpdateAccessRecords(workspace) andThen
+            deleteAttributes(workspace) andThen
+            insertAttributes(workspace) andThen
+            optimisticLockUpdate(workspaceRecord)
+      } map { _ => workspace }
+    }
+
+    private def insertOrUpdateAccessRecords(workspace: Workspace): WriteAction[Unit] = {
+      val id = UUID.fromString(workspace.workspaceId)
+      val accessRecords = workspace.accessLevels.map { case (accessLevel, group) => WorkspaceAccessRecord(id, group.groupName.value, accessLevel.toString, false) }
+      val realmAclRecords = workspace.realmACLs.map { case (accessLevel, group) => WorkspaceAccessRecord(id, group.groupName.value, accessLevel.toString, true) }
+      DBIO.seq((accessRecords ++ realmAclRecords).map { workspaceAccessQuery insertOrUpdate }.toSeq: _*)
+    }
+
+    private def deleteAttributes(workspace: Workspace): ReadWriteAction[Unit] = {
+      val workspaceId = UUID.fromString(workspace.workspaceId)
+      workspaceAttributes(findByIdQuery(workspaceId)).result.flatMap { wsIdAndAttributeRecords =>
+        val records = wsIdAndAttributeRecords.map { case (wsId, attrRec) => attrRec }
+        // clear existing attributes, any concurrency locking should be handled at the workspace level
+        val deleteActions = deleteWorkspaceAttributes(records)
+        DBIO.seq(deleteActions.toSeq:_*)
+      }
+    }
 
-          // insert attributes
-          val insertActions = workspace.attributes.flatMap { case (name, attribute) =>
-            attributeQuery.insertAttributeRecords(name, attribute, workspaceRecord.id).map(_.flatMap{ attrId =>
-              insertWorkspaceAttributeMapping(workspaceRecord, attrId)
-            })
-          }
+    private def insertAttributes(workspace: Workspace): ReadWriteAction[Unit] = {
+      val workspaceId = UUID.fromString(workspace.workspaceId)
+      val insertActions = workspace.attributes.flatMap { case (name, attribute) =>
+        attributeQuery.insertAttributeRecords(name, attribute, workspaceId).map(_.flatMap{ attrId =>
+          insertWorkspaceAttributeMapping(workspaceId, attrId)
+        })
+      }
+      DBIO.seq(insertActions.toSeq:_*)
+    }
 
-          DBIO.seq(deleteActions ++ insertActions:_*)
-        }
-      } map(_ => workspace)
+    private def optimisticLockUpdate(originalRec: WorkspaceRecord): ReadWriteAction[Int] = {
+      findByIdAndRecordVersionQuery(originalRec.id, originalRec.recordVersion) update originalRec.copy(recordVersion = originalRec.recordVersion + 1) map {
+        case 0 => throw new RawlsConcurrentModificationException(s"could not update $originalRec because its record version has changed")
```

Comment:
if, in production, we saw the corruption happen with relative frequency ... then this means we'll be throwing exceptions with relative frequency. What should the user experience be? Perhaps we need new stories to handle this exception in the UI. (I don't have good suggestions right now on what the UX should be ... just pointing out that exceptions may be scary to users)